### ISSUE TITLE
Add a new exclude syntax to be able to exclude a whole package from bundling 

### ISSUE THREAD

**Issue by izaera** [liferay/liferay-npm-build-tools#99]
This is needed to be able to stop transitive dependencies from being bundled in the finar JAR, when they are not needed or we want to import them, for example.

The syntax would be:

```json
{
    "exclude": {
        "metal": true
    }
}
```

Where `true` means `exclude it` and `false` would mean `keep it`.

**Comment by yuchi** [comment]
@izaera is there a place where you are gonna describe the new approach?

I’m having a hard time understanding how a dev could implement something similar to the image-editor, where a “main” bundle defines a dependnecy on a framework (or set of libraries) and the extensions inherit those dependencies.

**Comment by izaera** [comment]
It's still experimental: we may introduce new changes if we see we need them for the portal source/build structure.

Just as a quick summary the main idea has been isolating dependencies per OSGi bundle (a la webpack) and then letting users point isolated dependencies to different bundles by configuring an `imports` section in the `.npmbundlerrc`.

That way, the responsibility of deduplication is left to the user. It's inspired in vendor chunks of webpack, but it's not the same, of course. Webpack's main goal is to speed up build (more like C's static libraries) while we do it to be able to link modules during runtime (more similar to .DLL/.so).

**Comment by yuchi** [comment]
Why should a developer need to add a package to `exclude` since it is already in `import[parent-bundle]`?

Is there any case where this doesn’t happen? (you `import` but not `exclude`?)

**Comment by yuchi** [comment]
(BTW thank you, I’ll have a look at your `WIP`s)

**Comment by izaera** [comment]
There are two reasons for the need to add a package to `exclude` and `imports` section. 

The first one is technical: because the bundler is modular and based on plugins I have added the `config` section so that several plugins may share the same configuration between them. Unfortunately, the `exclude` section is not based on plugins, but applied _a priori_ by the bundler itself, so there's no way to share that configuration with the plugins.

Said that, we could change the bundler to make the exclude section a plugin based one and that way the configuration would be unique. My only concern is that, conceptually, exclusions and imports are not the same. I mean, you can import a package but still want to ship that package locally because they are different things.

Possible scenarios for that would be: 

* Shipping the local package so that, in case it cannot be imported, we can fall back to the local copy.
* Shipping the local package so that, even if we are importing it from somewhere else, someone can consume our package (this sounds quite useless, but who knows...).
* Any other we have not yet think of...

To sum up: I'm not sure what is best. Perhaps we can have both strategies and decide which one to use (being the unique config the default). We can also leave it like it is now because normally, you don't configure imports by hand, but using an automatic tool (that needs to be done). And that tool would take care of syncing the imports and the exclude section.

Whichever decission we make we can always change it in a following version (is not a decission that would make us stick to it uniquely in the future) so...

Opinions? @jbalsas @yuchi

**Comment by izaera** [comment]
OK. In a private conversation with @jbalsas we have decided to make unique configuration (imports => exclude) the default, and in the future, if it is needed provide a way to override it (in case someone wants to import but not exlude).

**Comment by izaera** [comment]
**Fiendly 📚 warning**

Just to make clear the decissions we are taking I'm pasting a huge philosophical thread (Watch out 😜!).

OK, so imagine we have a project with this dependency tree:

```
project
  -> dep@1.0.0
  -> transitive@1.0.0
    -> dep@2.0.0
```

If we run the new bundler (that namespaces dependencies with the project's name) through it, we get the following JAR:

```
package.json
  node_modules
    project$dep@1.0.0
    project$transitive@1.0.0
    project$dep@2.0.0
```

As you can see, there are two `dep` packages in the final JAR. This is because the loader will give `dep@1.0.0` to `project` and `dep@2.0.0` to `transitive@1.0.0`. This is in absence of any `imports`.

Now, suppose we add the following `imports` to `.npmbundlerrc`:

```
{
  "config": {
    "imports": {
      "other-project": {
        "dep": "^1.0.0"
      }
    }
  }
}
```

What should happen now when resolving `dep` for the `project` and `dep@1.0.0` packages (case A)? 

Furthermore, what would happen if there was only one `dep` because `transitive@1.0.0`'s semver constraints are the same as those of `project` (that would be `1.0.0`) and npm dedupes them (case B)?

There are several possibilities:

Case A:
1. `project` gets `dep@^1.0.0` from the external bundle `other-project` and `transitive@1.0.0` gets `dep@2.0.0` from `project`'s bundle.
2. Everybody (`project` and `transitive@1.0.0`) gets `dep@^1.0.0` from the external bundle `other-project` (note that `transitive@1.0.0` may fail because it will get a different version than the one it is expecting).

Case B:
1. `transitive@1.0.0` gets `dep@^1.0.0` from the external bundle `other-project`
2. `transitive@1.0.0` gets `dep@1.0.0` from its bundle (without importing it).

Looks like `1` for case `A` is the mathematically correct solution and, in that case, for coherence, we would choose solution `2` for case `B` because it would mean that the `imports` only affect the project's dependencies and not the `node_modules` packages dependencies. Of course we could also do `1` for case `B` but it would break simmetry with case `A`.

At the same time, we feel that `1` for `A` and `2` for `B` is counter-intuitive, because if you say you want to import something the expectation is that it is always imported, not just sometimes. 

Also, we think it is quite uncommon to try to import a package that it is also being used by an inner dependency in another version. The rationale behind imports is that you want to share some packages between different projects because someone is managing centralized versions of that packages so, why would you want to use a different version in one of your projects? In that case, it would be more rational to stop importing that dependency.

Thus, we will opt for `2` for case `A` and `1` for case `B`. That means that, once you import some package with certain version constraints, it will be used from any package inside your bundle (be it your project or a `node_modules` package) and with that version constraints. And then, the bundler will warn you if your configuration may lead to some package getting a different version than the one it is expecting because of your `imports` (and you decide what to do then).

Fortunately, this decission doesn't close the door to other solutions in the future as we can always modify the bundler in a compatible way and, using config, achieve other strategies of importing.
